src/python/pants/goal/run_tracker.py

```diff
@@ -492,8 +492,9 @@ def end(self):
     return 1 if outcome in [WorkUnit.FAILURE, WorkUnit.ABORTED] else 0
 
   def end_workunit(self, workunit):
-    self.report.end_workunit(workunit)
     path, duration, self_time, is_tool = workunit.end()
+    self.report.end_workunit(workunit)
+    workunit.close_outputs()
```

Comment:
I'm a little confused as to why we're separating out the closing of outputs from the workunit end. Could you explain that bit? I feel like it makes more sense for `workunit.end()` to handle that, but if it's preventing something from happening it'd be good to better understand that.

Reply:
This is so we don't close outputs that the report / reporters might attempt to read - such as in `report.end_workunit(workunit)` called on `l496`, which (among other things) calls `_notify()`:

https://github.com/pantsbuild/pants/blob/23bcfe910417c103c8144b083a56b8fed998ee9f/src/python/pants/reporting/report.py#L138-L142

Reply:
I see. So, does that mean that there'd be test failures without splitting end/close_outputs?

Reply:
Yup (`ValueError - seek of closed file`)